pyiceberg/partitioning.py

```diff
@@ -272,6 +272,60 @@ def assign_fresh_partition_spec_ids(spec: PartitionSpec, old_schema: Schema, fre
 T = TypeVar("T")
 
 
+class PartitionMap(Generic[T]):
+    _specs: dict[int, PartitionSpec]
+    _partition_maps: dict[int, dict[Record | None, T]]
+
+    def __init__(self, specs: dict[int, PartitionSpec]):
+        self._specs = specs
+        self._partition_maps = {}
+
+    def __len__(self) -> int:
+        """Return the length of the partition map.
+
+        Returns:
+            length of _partition_maps
+        """
+        return len(self.values())
+
+    def is_empty(self) -> bool:
+        return len(self._partition_maps.values()) == 0
```

Comment:
If we are following the [Java implementation](https://github.com/apache/iceberg/blob/026374b530eeea46429cef61a99ba9321e0587b6/core/src/main/java/org/apache/iceberg/util/PartitionMap.java#L74), I think we'd want to check that the list of values of each partition map is empty, instead of the current behavior, which checks that the length of `self._partition_maps` is 0.

```suggestion
        return all(len(partition_map) == 0 for partition_map in self._partition_maps.values())
```